getSF does not work for data with frequency selection and slow N axis

Given:
A system matrix saved like this:
* mdf.measurement.data = reshape(S_cmplx, K, C, J, N)
* mdf.measurement.isFastFrameAxis = Int8(0)
* mdf.measurement.frequencySelection = [Int64(indx) for indx in 1:n_total_freqs if frequency_selection_mask[indx]]
* mdf.acquisition.receiver.numSamplingPoints = (n_total_freqs - 1)*2 + 1
* ... other fields

where K is the reduced number of frequencies, i.e. K=sum(frequency_selection_mask). n_total_freqs is the original number of frequencies, i.e. n_total_freqs = length(frequency_selection_mask)

After saving and loading using
```julia
S, grid = getSF(sf)
```
an error is raised because `measDataTD` tries to apply irfft using the original number of sampling points to the data which is filtered. The problem would not appear, if rxNumSamplingPoints was equal K, but then `rowsToSubsampledRows` would not work as it expects it to contain the original number of frequencies ([here is the function which would break](https://github.com/MagneticParticleImaging/MPIFiles.jl/blob/5c985388beee3f0376eddff7d816cdfafe80fbe7/src/FrequencyFilter.jl#L209)). At the same time, there is no need to apply irfft and convert to the data to the time domain. The only transform required is `permutedims`.

```
ERROR: AssertionError: sz[d1] == d >> 1 + 1
Stacktrace:
  [1] brfft_output_size
    @ ~/.julia/packages/AbstractFFTs/4iQz5/src/definitions.jl:350 [inlined]
  [2] brfft_output_size
    @ ~/.julia/packages/AbstractFFTs/4iQz5/src/definitions.jl:347 [inlined]
  [3] plan_brfft(X::Array{ComplexF64, 4}, d::Int64, region::Int64; flags::UInt32, timelimit::Float64, num_threads::Nothing)
    @ FFTW ~/.julia/packages/FFTW/6nZei/src/fft.jl:869
  [4] plan_brfft
    @ ~/.julia/packages/FFTW/6nZei/src/fft.jl:859 [inlined]
  [5] #plan_irfft#27
    @ ~/.julia/packages/AbstractFFTs/4iQz5/src/definitions.jl:354 [inlined]
  [6] plan_irfft
    @ ~/.julia/packages/AbstractFFTs/4iQz5/src/definitions.jl:354 [inlined]
  [7] irfft(x::Array{ComplexF64, 4}, d::Int64, region::Int64)
    @ AbstractFFTs ~/.julia/packages/AbstractFFTs/4iQz5/src/definitions.jl:306
  [8] measDataTD(f::MDFFileV2{Array{ComplexF64, 4}}, frames::UnitRange{Int64}, periods::UnitRange{Int64}, receivers::UnitRange{Int64})
    @ MPIFiles ~/.julia/packages/MPIFiles/NVcPK/src/Derived.jl:141
  [9] measDataTD(f::MDFFileV2{Array{ComplexF64, 4}}, frames::UnitRange{Int64}, periods::UnitRange{Int64})
    @ MPIFiles ~/.julia/packages/MPIFiles/NVcPK/src/Derived.jl:132
 [10] measDataConv(::MDFFileV2{Array{ComplexF64, 4}}, ::UnitRange{Int64}, ::Vararg{Any})
    @ MPIFiles ~/.julia/packages/MPIFiles/NVcPK/src/Measurements.jl:4
 [11] measDataLowLevel(::MDFFileV2{Array{ComplexF64, 4}}, ::UnitRange{Int64}, ::Vararg{UnitRange{Int64}}; spectralLeakageCorrection::Bool)
    @ MPIFiles ~/.julia/packages/MPIFiles/NVcPK/src/Measurements.jl:106
 [12] measDataLowLevel
    @ ~/.julia/packages/MPIFiles/NVcPK/src/Measurements.jl:103 [inlined]
 [13] getAveragedMeasurements(f::MDFFileV2{Array{ComplexF64, 4}}; frames::UnitRange{Int64}, numAverages::Int64, periods::UnitRange{Int64}, averagePeriodsPerPatch::Bool, numPeriodAverages::Int64, fixDistortions::Bool, kargs::Base.Pairs{Symbol, Bool, Tuple{Symbol}, NamedTuple{(:spectralLeakageCorrection,), Tuple{Bool}}})
    @ MPIFiles ~/.julia/packages/MPIFiles/NVcPK/src/Measurements.jl:125
 [14] getMeasurements(f::MDFFileV2{Array{ComplexF64, 4}}, neglectBGFrames::Bool; frames::UnitRange{Int64}, bgCorrection::Bool, interpolateBG::Bool, tfCorrection::Bool, sortFrames::Bool, numAverages::Int64, numPeriodGrouping::Int64, kargs::Base.Pairs{Symbol, Integer, Tuple{Symbol, Symbol}, NamedTuple{(:spectralLeakageCorrection, :numPeriodAverages), Tuple{Bool, Int64}}})
    @ MPIFiles ~/.julia/packages/MPIFiles/NVcPK/src/Measurements.jl:197
 [15] getMeasurements
    @ ~/.julia/packages/MPIFiles/NVcPK/src/Measurements.jl:189 [inlined]
 [16] getMeasurementsFD(::MDFFileV2{Array{ComplexF64, 4}}; loadasreal::Bool, transposed::Bool, frequencies::Vector{CartesianIndex{2}}, tfCorrection::Bool, kargs::Base.Pairs{Symbol, Integer, NTuple{5, Symbol}, NamedTuple{(:sortFrames, :bgCorrection, :spectralLeakageCorrection, :numPeriodAverages, :numPeriodGrouping), Tuple{Bool, Bool, Bool, Int64, Int64}}})
    @ MPIFiles ~/.julia/packages/MPIFiles/NVcPK/src/Measurements.jl:326
 [17] getMeasurementsFD
    @ ~/.julia/packages/MPIFiles/NVcPK/src/Measurements.jl:322 [inlined]
 [18] getSystemMatrix(f::MDFFileV2{Array{ComplexF64, 4}}, frequencies::Vector{CartesianIndex{2}}; bgCorrection::Bool, loadasreal::Bool, tfCorrection::Bool, numPeriodAverages::Int64, numPeriodGrouping::Int64, kargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ MPIFiles ~/.julia/packages/MPIFiles/NVcPK/src/SystemMatrix.jl:26
 [19] getSystemMatrix
    @ ~/.julia/packages/MPIFiles/NVcPK/src/SystemMatrix.jl:13 [inlined]
 [20] getSF(bSF::MDFFileV2{Array{ComplexF64, 4}}, frequencies::Vector{CartesianIndex{2}}; returnasmatrix::Bool, procno::Int64, bgcorrection::Bool, bgCorrection::Bool, loadasreal::Bool, gridsize::Vector{Int64}, numPeriodAverages::Int64, numPeriodGrouping::Int64, fov::Vector{Float64}, center::Vector{Float64}, deadPixels::Nothing, kargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ MPIReco ~/.julia/packages/MPIReco/YsDsR/src/SystemMatrix/SystemMatrix.jl:189
 [21] getSF(bSF::MDFFileV2{Array{ComplexF64, 4}}, frequencies::Vector{CartesianIndex{2}})
    @ MPIReco ~/.julia/packages/MPIReco/YsDsR/src/SystemMatrix/SystemMatrix.jl:180
 [22] #getSF#132
    @ ~/.julia/packages/MPIReco/YsDsR/src/SystemMatrix/SystemMatrix.jl:138 [inlined]
 [23] getSF(f::MDFFileV2{Array{ComplexF64, 4}})
    @ MPIReco ~/.julia/packages/MPIReco/YsDsR/src/SystemMatrix/SystemMatrix.jl:138
```

--- a/src/FrequencyFilter.jl
+++ b/src/FrequencyFilter.jl
@@ -68,9 +68,11 @@ function filterFrequencies(f::MPIFile; SNRThresh=-1, minFreq=0,
   SNRAll = nothing
 
   if SNRThresh > 0 || numUsedFreqs > 0
+    SNR = zeros(eltype(SNRThresh), nFreq, nReceivers)
+
     SNRAll = calibSNR(f)
     if !isnothing(SNRAll)
-      SNR = Dict(CartesianIndex(freqs[i], j) => SNRAll[i, j, 1] for i = 1:length(freqs), j in recChannels)
+      SNR[freqs,:] = SNRAll[:,:,1]
     end
   end
 
@@ -145,7 +147,7 @@ function filterFrequenciesBySNRThresh!(indices, f::MPIFile, snrthresh; numPeriod
   SNR = getCalibSNR(f, numPeriodGrouping = numPeriodGrouping)
   return filterFrequenciesBySNRThresh!(indices, snrthresh, SNR)
 end
-filterFrequenciesBySNRThresh!(indices, SNRThresh, SNR) = filter!(x-> SNR[x] >= SNRThresh , indices)
+filterFrequenciesBySNRThresh!(indices, SNRThresh, SNR::Matrix) = filter!(x-> SNR[x] >= SNRThresh , indices)
 filterFrequenciesBySNRThresh!(indices, SNRThresh, SNR::Dict{CartesianIndex{2}, Float64}) = filter!(x-> get(SNR, x, 0.0) >= SNRThresh , indices)
 
 

--- a/src/FrequencyFilter.jl
+++ b/src/FrequencyFilter.jl
@@ -36,11 +36,15 @@ function filterFrequencies(f::MPIFile; SNRThresh=-1, minFreq=0,
   nFreq = rxNumFrequencies(f, numPeriodGrouping)
   nReceivers = rxNumChannels(f)
   nPeriods = 1 #acqNumPeriodsPerFrame(f)
-  freqIndices = collect(vec(CartesianIndices((nFreq, nReceivers))))
+  freqs = measIsFrequencySelection(f) ? measFrequencySelection(f) : 1:nFreq
 
-  filterFrequenciesBySelection!(freqIndices, f)
-  filterFrequenciesByChannel!(freqIndices, recChannels)
-  
+  if numPeriodGrouping == 1
+    freqIndices = vec([CartesianIndex{2}(i, j) for i in freqs, j in recChannels])
+  else
+    freqIndices = collect(vec(CartesianIndices((nFreq, nReceivers))))
+    filterFrequenciesBySelection!(freqIndices, f) 
+    filterFrequenciesByChannel!(freqIndices, recChannels)
+  end  
   if minFreq > 0
     filterFrequenciesByMinFreq!(freqIndices, f, minFreq)
   end
@@ -64,12 +68,9 @@ function filterFrequencies(f::MPIFile; SNRThresh=-1, minFreq=0,
   SNRAll = nothing
 
   if SNRThresh > 0 || numUsedFreqs > 0
-    SNR = zeros(nFreq, nReceivers)
-    idx = measIsFrequencySelection(f) ? measFrequencySelection(f) : idx = 1:nFreq
-
     SNRAll = calibSNR(f)
     if !isnothing(SNRAll)
-      SNR[idx,:] = SNRAll[:,:,1]
+      SNR = Dict(CartesianIndex(freqs[i], j) => SNRAll[i, j, 1] for i = 1:length(freqs), j in recChannels)
     end
   end
 
@@ -86,42 +87,42 @@ function filterFrequencies(f::MPIFile; SNRThresh=-1, minFreq=0,
     filterFrequenciesByStepsize!(freqIndices, stepsize)
   end
 
-  return freqIndices
+  return collect(vec(freqIndices))
 end
 
 export filterFrequenciesBySelection!
-function filterFrequenciesBySelection!(indices::Vector{CartesianIndex{2}}, f::MPIFile)
+function filterFrequenciesBySelection!(indices, f::MPIFile)
   if measIsFrequencySelection(f)
     return filterFrequenciesBySelection!(indices, measFrequencySelection(f))
   end
 end
-filterFrequenciesBySelection!(indices::Vector{CartesianIndex{2}}, selection::Vector{Int64}) = filter!(x -> in(x[1], selection), indices)
+filterFrequenciesBySelection!(indices, selection::Vector{Int64}, sorted = issorted(selection)) = filter!(x -> sorted ? insorted(x[1], selection) : in(x[1], selection), indices)
 
 export filterFrequenciesByChannel!
-filterFrequenciesByChannel!(indices::Vector{CartesianIndex{2}}, channels) = filter!(x-> in(x[2], channels), indices)
+filterFrequenciesByChannel!(indices, channels) = filter!(x-> in(x[2], channels), indices)
 
 export filterFrequenciesByMinFreq!
-function filterFrequenciesByMinFreq!(indices::Vector{CartesianIndex{2}}, f::MPIFile, minFreq; numPeriodGrouping = 1)
+function filterFrequenciesByMinFreq!(indices, f::MPIFile, minFreq; numPeriodGrouping = 1)
   nFreq = rxNumFrequencies(f, numPeriodGrouping)
   minIdx = floor(Int, minFreq / rxBandwidth(f) * (nFreq-1) ) + 1
   return filterFrequenciesByMinIdx!(indices, minIdx)
 end
-filterFrequenciesByMinIdx!(indices::Vector{CartesianIndex{2}}, minIdx) = minIdx > 0 ?  filter!(x -> x[1] > minIdx, indices) : indices 
+filterFrequenciesByMinIdx!(indices, minIdx) = minIdx > 0 ?  filter!(x -> x[1] > minIdx, indices) : indices 
 
 export filterFrequenciesByMaxFreq!
-function filterFrequenciesByMaxFreq!(indices::Vector{CartesianIndex{2}}, f::MPIFile, maxFreq; numPeriodGrouping = 1)
+function filterFrequenciesByMaxFreq!(indices, f::MPIFile, maxFreq; numPeriodGrouping = 1)
   nFreq = rxNumFrequencies(f, numPeriodGrouping)
   maxIdx = ceil(Int, maxFreq / rxBandwidth(f) * (nFreq-1) ) + 1
   return filterFrequenciesByMaxIdx!(indices, maxIdx)
 end
-filterFrequenciesByMaxIdx!(indices::Vector{CartesianIndex{2}}, maxIdx) = filter!(x-> x[1] < maxIdx, indices)
+filterFrequenciesByMaxIdx!(indices, maxIdx) = filter!(x-> x[1] < maxIdx, indices)
 
 export filterFrequenciesByMaxMixingOrder!
-filterFrequenciesByMaxMixingOrder!(indices::Vector{CartesianIndex{2}}, maxMixingOrder, f::MPIFile) = filterFrequenciesByMaxMixingOrder!(indices, maxMixingOrder, mixingFactors(f))
-filterFrequenciesByMaxMixingOrder!(indices::Vector{CartesianIndex{2}}, maxMixingOrder, mf::Matrix) = filter!(x-> mf[x[1], 4] <= maxMixingOrder, indices)
+filterFrequenciesByMaxMixingOrder!(indices, maxMixingOrder, f::MPIFile) = filterFrequenciesByMaxMixingOrder!(indices, maxMixingOrder, mixingFactors(f))
+filterFrequenciesByMaxMixingOrder!(indices, maxMixingOrder, mf::Matrix) = filter!(x-> mf[x[1], 4] <= maxMixingOrder, indices)
 
 export filterFrequenciesByNumSidebandFreqs!
-function filterFrequenciesByNumSidebandFreqs!(indices::Vector{CartesianIndex{2}}, numSidebandFreqs::Int64, f::MPIFile; numPeriodGrouping = 1)
+function filterFrequenciesByNumSidebandFreqs!(indices, numSidebandFreqs::Int64, f::MPIFile; numPeriodGrouping = 1)
   # https://en.wikipedia.org/wiki/Sideband
   # Because of period grouping we have more frequencies than in original data
 
@@ -140,17 +141,19 @@ function filterFrequenciesByNumSidebandFreqs!(indices::Vector{CartesianIndex{2}}
 end
 
 export filterFrequenciesBySNRThresh!
-function filterFrequenciesBySNRThresh!(indices::Vector{CartesianIndex{2}}, f::MPIFile, snrthresh; numPeriodGrouping = 1)
+function filterFrequenciesBySNRThresh!(indices, f::MPIFile, snrthresh; numPeriodGrouping = 1)
   SNR = getCalibSNR(f, numPeriodGrouping = numPeriodGrouping)
   return filterFrequenciesBySNRThresh!(indices, snrthresh, SNR)
 end
-filterFrequenciesBySNRThresh!(indices::Vector{CartesianIndex{2}}, SNRThresh, SNR) = filter!(x-> SNR[x] >= SNRThresh , indices)
+filterFrequenciesBySNRThresh!(indices, SNRThresh, SNR) = filter!(x-> SNR[x] >= SNRThresh , indices)
+filterFrequenciesBySNRThresh!(indices, SNRThresh, SNR::Dict{CartesianIndex{2}, Float64}) = filter!(x-> get(SNR, x, 0.0) >= SNRThresh , indices)
+
 
 export filterFrequenciesByNumUsedFrequencies!
-function filterFrequenciesByNumUsedFrequencies!(indices::Vector{CartesianIndex{2}}, f::MPIFile, maxFreq)
+function filterFrequenciesByNumUsedFrequencies!(indices, f::MPIFile, maxFreq)
   error("Not implemented")
 end
-filterFrequenciesByNumUsedFrequencies!(indices::Vector{CartesianIndex{2}}, maxIdx) = error("not implemented")
+filterFrequenciesByNumUsedFrequencies!(indices, maxIdx) = error("not implemented")
 #=
     numFreqsAlreadyFalse = sum(!freqMask)
     numFreqsFalse = round(Int,length(freqMask)* (1-numUsedFreqs))
@@ -168,12 +171,12 @@ filterFrequenciesByNumUsedFrequencies!(indices::Vector{CartesianIndex{2}}, maxId
 =#
 
 export filterFrequenciesByStepsize!
-function filterFrequenciesByStepsize!(indices::Vector{CartesianIndex{2}}, stepsize)
+function filterFrequenciesByStepsize!(indices, stepsize)
   stepIndices = 1:stepsize:maximum(map(x -> x[1], indices))
   filter!(x -> in(x[1], stepIndices), indices)
 end
 
-function sortFrequencies(indices::Vector{CartesianIndex{2}}, f::MPIFile; numPeriodGrouping = 1, stepsize = 1, sortBySNR = false, sortByMixFactors = false)
+function sortFrequencies(indices, f::MPIFile; numPeriodGrouping = 1, stepsize = 1, sortBySNR = false, sortByMixFactors = false)
   if sortBySNR && !sortByMixFactors
     indices = sortFrequenciesBySNR(indices, f, numPeriodGrouping = numPeriodGrouping, stepsize = stepsize)
   elseif !sortBySNR && sortByMixFactors
@@ -185,11 +188,11 @@ function sortFrequencies(indices::Vector{CartesianIndex{2}}, f::MPIFile; numPeri
 end
 
 export sortFrequenciesBySNR
-function sortFrequenciesBySNR(indices::Vector{CartesianIndex{2}}, f::MPIFile; numPeriodGrouping = 1, stepsize = 1)
+function sortFrequenciesBySNR(indices, f::MPIFile; numPeriodGrouping = 1, stepsize = 1)
   SNR = getCalibSNR(f, numPeriodGrouping = numPeriodGrouping, stepsize = stepsize)
   sortFrequenciesBySNR(indices, SNR)
 end
-sortFrequenciesBySNR(indices::Vector{CartesianIndex{2}}, SNR::AbstractArray) = indices[reverse(sortperm(SNR[indices]),dims=1)]
+sortFrequenciesBySNR(indices, SNR::AbstractArray) = indices[reverse(sortperm(SNR[indices]),dims=1)]
 
 export sortFrequenciesByMixFactors
 function sortFrequenciesByMixFactors(indices, f::MPIFile; numPeriodGrouping = 1)
@@ -203,7 +206,7 @@ function sortFrequenciesByMixFactors(indices, f::MPIFile; numPeriodGrouping = 1)
   end
   sortFrequenciesByMixFactors(indices, mfNorm)
 end
-sortFrequenciesByMixFactors(indices::Vector{CartesianIndex{2}}, mfNorm::AbstractArray) = indices[sortperm(mfNorm[indices])]
+sortFrequenciesByMixFactors(indices, mfNorm::AbstractArray) = indices[sortperm(mfNorm[indices])]
 
 
 function rowsToSubsampledRows(f::MPIFile, rows)

--- a/src/FrequencyFilter.jl
+++ b/src/FrequencyFilter.jl
@@ -143,12 +143,12 @@ function filterFrequenciesByNumSidebandFreqs!(indices, numSidebandFreqs::Int64, 
 end
 
 export filterFrequenciesBySNRThresh!
-function filterFrequenciesBySNRThresh!(indices, f::MPIFile, snrthresh; numPeriodGrouping = 1)
-  SNR = getCalibSNR(f, numPeriodGrouping = numPeriodGrouping)
-  return filterFrequenciesBySNRThresh!(indices, snrthresh, SNR)
+function filterFrequenciesBySNRThresh!(indices, f::MPIFile, snrthresh::T; numPeriodGrouping = 1) where T <: Real
+  snr = getCalibSNR(f, numPeriodGrouping = numPeriodGrouping)
+  return filterFrequenciesBySNRThresh!(indices, snrthresh, snr)
 end
-filterFrequenciesBySNRThresh!(indices, SNRThresh, SNR::Matrix) = filter!(x-> SNR[x] >= SNRThresh , indices)
-filterFrequenciesBySNRThresh!(indices, SNRThresh, SNR::Dict{CartesianIndex{2}, Float64}) = filter!(x-> get(SNR, x, 0.0) >= SNRThresh , indices)
+filterFrequenciesBySNRThresh!(indices, snrthresh::T, snr::Matrix) where T <: Real = filter!(x-> snr[x] >= snrthresh , indices)
+filterFrequenciesBySNRThresh!(indices, snrthresh::T, snr::Dict{CartesianIndex{2}, Float64}) where T <: Real = filter!(x-> get(snr, x, 0.0) >= snrthresh , indices)
 
 
 export filterFrequenciesByNumUsedFrequencies!

--- a/src/FrequencyFilter.jl
+++ b/src/FrequencyFilter.jl
@@ -7,7 +7,7 @@ function getCalibSNR(f::MPIFile; numPeriodGrouping = 1, stepsize = 1)
   idx = measIsFrequencySelection(f) ? measFrequencySelection(f) : idx = 1:nFreq
 
   SNRAll = calibSNR(f)
-  if SNRAll != nothing
+  if !isnothing(SNRAll)
     SNR[idx,:] = SNRAll[:,:,1]
   end
 
@@ -68,11 +68,11 @@ function filterFrequencies(f::MPIFile; SNRThresh=-1, minFreq=0,
   SNRAll = nothing
 
   if SNRThresh > 0 || numUsedFreqs > 0
-    SNR = zeros(eltype(SNRThresh), nFreq, nReceivers)
+    SNR = zeros(Float64, nFreq, nReceivers)
 
     SNRAll = calibSNR(f)
     if !isnothing(SNRAll)
-      SNR[freqs,:] = SNRAll[:,:,1]
+      SNR[freqs, :] = SNRAll[:, :, 1]
     end
   end
 
@@ -83,7 +83,6 @@ function filterFrequencies(f::MPIFile; SNRThresh=-1, minFreq=0,
   elseif numUsedFreqs > 0 && SNRThresh > 0
     error("It is not possible to use SNRThresh and SNRFactorUsedFreq similtaneously")
   end
-
 
   if stepsize > 1
     filterFrequenciesByStepsize!(freqIndices, stepsize)
@@ -144,11 +143,11 @@ end
 
 export filterFrequenciesBySNRThresh!
 function filterFrequenciesBySNRThresh!(indices, f::MPIFile, snrthresh::T; numPeriodGrouping = 1) where T <: Real
-  snr = getCalibSNR(f, numPeriodGrouping = numPeriodGrouping)
-  return filterFrequenciesBySNRThresh!(indices, snrthresh, snr)
+  SNR = getCalibSNR(f, numPeriodGrouping = numPeriodGrouping)
+  return filterFrequenciesBySNRThresh!(indices, snrthresh, SNR)
 end
-filterFrequenciesBySNRThresh!(indices, snrthresh::T, snr::Matrix) where T <: Real = filter!(x-> snr[x] >= snrthresh , indices)
-filterFrequenciesBySNRThresh!(indices, snrthresh::T, snr::Dict{CartesianIndex{2}, Float64}) where T <: Real = filter!(x-> get(snr, x, 0.0) >= snrthresh , indices)
+filterFrequenciesBySNRThresh!(indices, SNRThresh::T, SNR::Matrix) where T <: Real = filter!(x-> SNR[x] >= SNRThresh , indices)
+filterFrequenciesBySNRThresh!(indices, SNRThresh::T, SNR::Dict{CartesianIndex{2}, Float64}) where T <: Real = filter!(x-> get(SNR, x, 0.0) >= SNRThresh , indices)
 
 
 export filterFrequenciesByNumUsedFrequencies!

--- a/src/FrequencyFilter.jl
+++ b/src/FrequencyFilter.jl
@@ -101,7 +101,7 @@ end
 filterFrequenciesBySelection!(indices, selection::Vector{Int64}, sorted = issorted(selection)) = filter!(x -> sorted ? insorted(x[1], selection) : in(x[1], selection), indices)
 
 export filterFrequenciesByChannel!
-filterFrequenciesByChannel!(indices, channels) = filter!(x-> in(x[2], channels), indices)
+filterFrequenciesByChannel!(indices, channels, sorted = issorted(channels)) = filter!(x-> sorted ? insorted(x[2], channels) : in(x[2], channels), indices)
 
 export filterFrequenciesByMinFreq!
 function filterFrequenciesByMinFreq!(indices, f::MPIFile, minFreq; numPeriodGrouping = 1)

--- a/src/FrequencyFilter.jl
+++ b/src/FrequencyFilter.jl
@@ -175,7 +175,7 @@ filterFrequenciesByNumUsedFrequencies!(indices, maxIdx) = error("not implemented
 export filterFrequenciesByStepsize!
 function filterFrequenciesByStepsize!(indices, stepsize)
   stepIndices = 1:stepsize:maximum(map(x -> x[1], indices))
-  filter!(x -> in(x[1], stepIndices), indices)
+  filter!(x -> insorted(x[1], stepIndices), indices)
 end
 
 function sortFrequencies(indices, f::MPIFile; numPeriodGrouping = 1, stepsize = 1, sortBySNR = false, sortByMixFactors = false)

--- a/src/FrequencyFilter.jl
+++ b/src/FrequencyFilter.jl
@@ -215,7 +215,7 @@ function rowsToSubsampledRows(f::MPIFile, rows)
     tmp = Array{Union{Nothing, CartesianIndex{2}}}(undef, rxNumFrequencies(f), rxNumChannels(f))
     idxAvailable = measFrequencySelection(f)
     for (i, idx) in enumerate(idxAvailable)
-      for d=1:rxNumChannels(f)
+      for d=1:size(tmp, 2)
         tmp[idx, d] = CartesianIndex(i, d)
       end
     end

--- a/src/MDFCommon.jl
+++ b/src/MDFCommon.jl
@@ -25,14 +25,17 @@ function measDataTDPeriods(f::Union{MDFFileV2, MDFv2InMemory}, periods=1:acqNumP
 end
 
 function systemMatrix(f::Union{MDFFileV2, MDFv2InMemory}, rows, bgCorrection=true)
-  if !experimentHasMeasurement(f) || !measIsFastFrameAxis(f) ||
-    !measIsFourierTransformed(f)
+  if !experimentHasMeasurement(f) || !measIsFourierTransformed(f)
     return nothing
   end
 
   rows_ = rowsToSubsampledRows(f, rows)
 
   data_ = measDataRaw(f)
+
+  if !measIsFastFrameAxis(f)
+    data_ = permutedims(data_, [4, 1, 2, 3])
+  end
 
   data_ = data_[:, rows_, :]
   data = reshape(data_, Val(2))

--- a/src/SystemMatrix.jl
+++ b/src/SystemMatrix.jl
@@ -19,7 +19,7 @@ function getSystemMatrix(f::MPIFile, frequencies=nothing;
     frequencies = filterFrequencies(f)
   end
 
-  if measIsFastFrameAxis(f) && measIsFourierTransformed(f)
+  if measIsFourierTransformed(f) && numPeriodGrouping == 1
     # This is the fast path if the SM lays on disk in the required format
     data = systemMatrix(f, frequencies, bgCorrection)
   else

--- a/test/General.jl
+++ b/test/General.jl
@@ -165,27 +165,27 @@
       #@test size(filterFrequencies(sm, numUsedFreqs = 100)) == (100,) # not working
 
       @test size(getSystemMatrix(sm,collect(vec(CartesianIndices((10, 1)))))) == (1936,10)
-      @test size(getSystemMatrix(sm,collect(vec(CartesianIndices((10, 1)))),loadasreal=true)) == (1936,20)
-      @test size(getSystemMatrix(sm,collect(vec(CartesianIndices((10, 1)))),bgCorrection=true)) == (1936,10)
+      @test size(getSystemMatrix(sm,collect(vec(CartesianIndices((10, 1)))), loadasreal=true)) == (1936,20)
+      @test size(getSystemMatrix(sm,collect(vec(CartesianIndices((10, 1)))), bgCorrection=true)) == (1936,10)
       # test on the data level if the conversion was successful
       SNRThresh = 2
-      freq = filterFrequencies(smBruker,SNRThresh=SNRThresh)
-      SBruker = getSystemMatrix(smBruker,frequencies=freq)
-      S = getSystemMatrix(sm,frequencies=freq)
+      freq = filterFrequencies(smBruker, SNRThresh=SNRThresh)
+      SBruker = getSystemMatrix(smBruker, frequencies=freq)
+      S = getSystemMatrix(sm, frequencies=freq)
 
-      relativeDeviation = zeros(Float32,length(freq))
+      relativeDeviation = zeros(Float32, length(freq))
       for f in 1:length(map(f->f[1], freq))
-        relativeDeviation[f] = norm(SBruker[:,f]-S[:,f])/norm(SBruker[:,f])
+        relativeDeviation[f] = norm(SBruker[:,f]-S[:,f]) / norm(SBruker[:,f])
       end
       # test if relative deviation for most of the frequency components is below 0.003
-      @test quantile(relativeDeviation,0.95)<0.003
+      @test quantile(relativeDeviation, 0.95) < 0.003
     end
 
     # Next test checks if the cached system matrix is the same as the one loaded
     # from the raw data
     S_loadedfromraw = getMeasurementsFD(smBrukerPretendToBeMeas,
-          frames=1:acqNumFGFrames(smBrukerPretendToBeMeas),sortFrames=true,
-          spectralLeakageCorrection=false,transposed=true,tfCorrection=false)
+          frames=1:acqNumFGFrames(smBrukerPretendToBeMeas), sortFrames=true,
+          spectralLeakageCorrection=false, transposed=true, tfCorrection=false)
 
     S_loadedfromproc = systemMatrix(smBruker)
 
@@ -232,7 +232,7 @@
       relativeDeviation[f] = norm(S1[:,f]-S2[:,f])/norm(S2[:,f])
     end
     # test if relative deviation for most of the frequency components is below 0.003
-    @test quantile(relativeDeviation,0.95)<0.14
+    @test quantile(relativeDeviation, 0.95)<0.14
 
 
     # Calibration file 1D
@@ -253,7 +253,7 @@
       @info "Test $sm"
 
       @test size( systemMatrixWithBG(sm) ) == (67,52,3,1)
-      @test size( systemMatrix(sm,collect(vec(CartesianIndices((10, 1))))) ) == (60,10)
+      @test size( systemMatrix(sm, collect(vec(CartesianIndices((10, 1))))) ) == (60,10)
       @test size( systemMatrix(sm) ) == (60,52,3,1)
 
       @test size(calibSNR(sm)) == (52, 3, 1)

